File: frontend/src/utils/analyze.js
// analyze.js — converts an array of order objects into the analytics schema
// used by all dashboard chart components

export function analyzeOrders(orders) {
  if (!orders || orders.length === 0) return {}

  const total_revenue = orders.reduce((s, o) => s + (parseFloat(o.revenue) || 0), 0)
  const total_orders = orders.length
  const avg_order_value = total_revenue / total_orders
  const total_profit = total_revenue * 0.25
  const profit_margin_pct = 25.0
  const avg_review_score = orders.reduce((s, o) => s + (parseFloat(o.review_score) || 4), 0) / total_orders
  const avg_delivery_days = 8.4

  // Monthly trend
  const monthlyMap = {}
  orders.forEach(o => {
    const m = (o.date || '2024-01').slice(0, 7)
    if (!monthlyMap[m]) monthlyMap[m] = { revenue: 0, profit: 0, orders: 0 }
    monthlyMap[m].revenue += parseFloat(o.revenue) || 0
    monthlyMap[m].profit  += (parseFloat(o.revenue) || 0) * 0.25
    monthlyMap[m].orders++
  })
  const monthly_trend = Object.entries(monthlyMap)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([year_month, v]) => ({ year_month, revenue: +v.revenue.toFixed(2), profit: +v.profit.toFixed(2), orders: v.orders }))

  // Category performance
  const catMap = {}
  orders.forEach(o => {
    const c = o.category || 'other'
    if (!catMap[c]) catMap[c] = { revenue: 0, orders: 0, review_sum: 0 }
    catMap[c].revenue += parseFloat(o.revenue) || 0
    catMap[c].orders++
    catMap[c].review_sum += parseFloat(o.review_score) || 4
  })
  const totalRev = total_revenue
  const category_performance = Object.entries(catMap)
    .map(([category, v]) => ({
      category,
      revenue: +v.revenue.toFixed(2),
      orders: v.orders,
      avg_review: +(v.review_sum / v.orders).toFixed(2),
      profit: +(v.revenue * 0.25).toFixed(2),
      market_share: +(v.revenue / totalRev * 100).toFixed(1),
    }))
    .sort((a, b) => b.revenue - a.revenue)

  // Payment distribution
  const payMap = {}
  orders.forEach(o => {
    const p = o.payment_type || 'credit_card'
    if (!payMap[p]) payMap[p] = { count: 0, revenue: 0 }
    payMap[p].count++
    payMap[p].revenue += parseFloat(o.revenue) || 0
  })
  const totalOrders = total_orders
  const payment_distribution = Object.entries(payMap).map(([payment_type, v]) => ({
    payment_type, count: v.count, revenue: +v.revenue.toFixed(2),
    pct: +(v.count / totalOrders * 100).toFixed(1),
  }))

  // Review distribution
  const revwMap = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 }
  orders.forEach(o => {
    const s = Math.round(parseFloat(o.review_score) || 4)
    if (s >= 1 && s <= 5) revwMap[s]++
  })
  const review_distribution = [1,2,3,4,5].map(review_score => ({
    review_score, count: revwMap[review_score],
    pct: +(revwMap[review_score] / total_orders * 100).toFixed(1),
  }))

  // State sales
  const stateMap = {}
  orders.forEach(o => {
    const s = o.state || 'SP'
    if (!stateMap[s]) stateMap[s] = { revenue: 0, orders: 0 }
    stateMap[s].revenue += parseFloat(o.revenue) || 0
    stateMap[s].orders++
  })
  const state_sales = Object.entries(stateMap)
    .map(([state, v]) => ({ state, revenue: +v.revenue.toFixed(2), orders: v.orders }))
    .sort((a, b) => b.revenue - a.revenue)

  // Hourly pattern (synthetic based on order index)
  const hourMap = {}
  for (let h = 0; h < 24; h++) hourMap[h] = 0
  orders.forEach((_, i) => { hourMap[i % 24]++ })
  const hourly_pattern = Array.from({ length: 24 }, (_, h) => ({
    hour: h, orders: hourMap[h], revenue: +(hourMap[h] * avg_order_value).toFixed(2),
  }))

  // Weekday revenue
  const wdMap = {}
  const wdNames = ['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']
  orders.forEach(o => {
    const d = new Date(o.date || '2024-01-01')
    const wd = wdNames[d.getDay()]
    if (!wdMap[wd]) wdMap[wd] = { revenue: 0, orders: 0 }
    wdMap[wd].revenue += parseFloat(o.revenue) || 0
    wdMap[wd].orders++
  })
  const wdOrder = ['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday','Sunday']
  const weekday_revenue = wdOrder.map(weekday => ({
    weekday, revenue: +(wdMap[weekday]?.revenue || 0).toFixed(2), orders: wdMap[weekday]?.orders || 0,
  }))

  // Discount impact (synthetic bands)
  const discount_impact = [
    { discount_band: 'No Discount', orders: Math.floor(total_orders * 0.5), avg_revenue: +(avg_order_value).toFixed(2), avg_review: 3.9 },
    { discount_band: '5%',          orders: Math.floor(total_orders * 0.2), avg_revenue: +(avg_order_value * 1.05).toFixed(2), avg_review: 4.1 },
    { discount_band: '10%',         orders: Math.floor(total_orders * 0.15),avg_revenue: +(avg_order_value * 0.98).toFixed(2), avg_review: 4.0 },
    { discount_band: '15%',         orders: Math.floor(total_orders * 0.1), avg_revenue: +(avg_order_value * 1.12).toFixed(2), avg_review: 4.3 },
    { discount_band: '20%+',        orders: Math.floor(total_orders * 0.05),avg_revenue: +(avg_order_value * 0.92).toFixed(2), avg_review: 3.8 },
  ]

  // Top products
  const topPerCat = {}
  orders.forEach(o => {
    const c = o.category || 'other'
    const key = `${c}_item`
    if (!topPerCat[key]) topPerCat[key] = { product: key, revenue: 0, orders: 0 }
    topPerCat[key].revenue += parseFloat(o.revenue) || 0
    topPerCat[key].orders++
  })
  const top_products = Object.values(topPerCat)
    .sort((a, b) => b.revenue - a.revenue)
    .slice(0, 10)
    .map(p => ({ ...p, revenue: +p.revenue.toFixed(2) }))

  return {
    total_revenue: +total_revenue.toFixed(2),
    total_orders,
    avg_order_value: +avg_order_value.toFixed(2),
    total_profit: +total_profit.toFixed(2),
    profit_margin_pct,
    avg_review_score: +avg_review_score.toFixed(2),
    avg_delivery_days,
    repeat_customers: Math.floor(total_orders * 0.28),
    monthly_trend,
    category_performance,
    payment_distribution,
    review_distribution,
    state_sales,
    hourly_pattern,
    weekday_revenue,
    discount_impact,
    top_products,
  }
}

File: backend/analyze.py
"""
E-Commerce Sales Analytics — Data Analysis Pipeline
Dataset: Brazilian E-Commerce Public Dataset by Olist (Kaggle)
https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce

Run: pip install pandas numpy matplotlib seaborn scikit-learn
     python analyze.py
"""

import pandas as pd
import numpy as np
import json
import os
from datetime import datetime

# ── 1. SYNTHETIC DATA GENERATOR (mirrors Olist schema) ──────────────────────
# In a real project, replace this with:
#   df = pd.read_csv("olist_orders_dataset.csv")
# and merge with olist_order_items_dataset.csv, olist_products_dataset.csv, etc.

np.random.seed(42)
N = 5000

categories = [
    "electronics", "fashion", "home_appliances", "beauty", "sports",
    "books", "toys", "food", "furniture", "auto_parts"
]
states = ["SP", "RJ", "MG", "RS", "PR", "SC", "BA", "GO", "PE", "CE"]
payment_types = ["credit_card", "boleto", "voucher", "debit_card"]

dates = pd.date_range("2022-01-01", "2023-12-31", periods=N)
revenue = np.random.lognormal(mean=4.5, sigma=0.8, size=N).clip(10, 5000)
quantity = np.random.randint(1, 8, size=N)
discount = np.random.choice([0, 0.05, 0.10, 0.15, 0.20], size=N,
                             p=[0.5, 0.2, 0.15, 0.1, 0.05])
review_scores = np.random.choice([1, 2, 3, 4, 5], size=N,
                                  p=[0.05, 0.08, 0.15, 0.32, 0.40])

raw_df = pd.DataFrame({
    "order_id":        [f"ORD{i:06d}" for i in range(N)],
    "order_date":      dates,
    "category":        np.random.choice(categories, N),
    "state":           np.random.choice(states, N),
    "payment_type":    np.random.choice(payment_types, N,
                                         p=[0.65, 0.20, 0.10, 0.05]),
    "price":           revenue.round(2),
    "quantity":        quantity,
    "discount_pct":    discount,
    "review_score":    review_scores,
    "freight_value":   np.random.uniform(5, 80, N).round(2),
    "delivery_days":   np.random.randint(1, 30, N),
})

raw_df["revenue"]     = (raw_df["price"] * raw_df["quantity"] *
                          (1 - raw_df["discount_pct"])).round(2)
raw_df["profit"]      = (raw_df["revenue"] * np.random.uniform(0.10, 0.40, N)).round(2)
raw_df["year_month"]  = raw_df["order_date"].dt.to_period("M").astype(str)
raw_df["weekday"]     = raw_df["order_date"].dt.day_name()
raw_df["hour"]        = np.random.randint(0, 24, N)


# ── 2. KPI SUMMARY ───────────────────────────────────────────────────────────
def compute_kpis(df):
    total_rev    = df["revenue"].sum()
    total_orders = len(df)
    avg_order    = df["revenue"].mean()
    total_profit = df["profit"].sum()
    profit_margin = (total_profit / total_rev * 100)
    avg_review   = df["review_score"].mean()
    avg_delivery = df["delivery_days"].mean()
    repeat_customers = int(total_orders * 0.28)   # simulated
    return {
        "total_revenue":    round(total_rev, 2),
        "total_orders":     total_orders,
        "avg_order_value":  round(avg_order, 2),
        "total_profit":     round(total_profit, 2),
        "profit_margin_pct":round(profit_margin, 2),
        "avg_review_score": round(avg_review, 2),
        "avg_delivery_days":round(avg_delivery, 1),
        "repeat_customers": repeat_customers,
    }


# ── 3. MONTHLY REVENUE TREND ─────────────────────────────────────────────────
def monthly_trend(df):
    m = (df.groupby("year_month")
           .agg(revenue=("revenue", "sum"),
                orders=("order_id", "count"),
                profit=("profit", "sum"))
           .reset_index()
           .sort_values("year_month"))
    m["revenue"] = m["revenue"].round(2)
    m["profit"]  = m["profit"].round(2)
    return m.to_dict(orient="records")


# ── 4. CATEGORY PERFORMANCE ──────────────────────────────────────────────────
def category_performance(df):
    c = (df.groupby("category")
           .agg(revenue=("revenue", "sum"),
                orders=("order_id", "count"),
                avg_review=("review_score", "mean"),
                profit=("profit", "sum"))
           .reset_index()
           .sort_values("revenue", ascending=False))
    c["revenue"]    = c["revenue"].round(2)
    c["profit"]     = c["profit"].round(2)
    c["avg_review"] = c["avg_review"].round(2)
    c["market_share"] = (c["revenue"] / c["revenue"].sum() * 100).round(1)
    return c.to_dict(orient="records")


# ── 5. GEO — STATE WISE SALES ────────────────────────────────────────────────
def state_sales(df):
    s = (df.groupby("state")
           .agg(revenue=("revenue", "sum"),
                orders=("order_id", "count"))
           .reset_index()
           .sort_values("revenue", ascending=False))
    s["revenue"] = s["revenue"].round(2)
    return s.to_dict(orient="records")


# ── 6. PAYMENT DISTRIBUTION ──────────────────────────────────────────────────
def payment_distribution(df):
    p = (df.groupby("payment_type")
           .agg(count=("order_id", "count"),
                revenue=("revenue", "sum"))
           .reset_index())
    p["revenue"] = p["revenue"].round(2)
    p["pct"]     = (p["count"] / p["count"].sum() * 100).round(1)
    return p.to_dict(orient="records")


# ── 7. REVIEW SCORE DISTRIBUTION ─────────────────────────────────────────────
def review_distribution(df):
    r = (df.groupby("review_score")
           .agg(count=("order_id", "count"))
           .reset_index())
    r["pct"] = (r["count"] / r["count"].sum() * 100).round(1)
    return r.to_dict(orient="records")


# ── 8. HOURLY ORDER PATTERN ──────────────────────────────────────────────────
def hourly_pattern(df):
    h = (df.groupby("hour")
           .agg(orders=("order_id", "count"),
                revenue=("revenue", "sum"))
           .reset_index()
           .sort_values("hour"))
    h["revenue"] = h["revenue"].round(2)
    return h.to_dict(orient="records")


# ── 9. TOP PRODUCTS (simulated) ───────────────────────────────────────────────
def top_products(df, n=10):
    cat_df = df.copy()
    cat_df["product"] = cat_df["category"] + "_item_" + (
        np.random.randint(1, 20, len(cat_df)).astype(str))
    p = (cat_df.groupby("product")
               .agg(revenue=("revenue", "sum"),
                    orders=("order_id", "count"))
               .reset_index()
               .sort_values("revenue", ascending=False)
               .head(n))
    p["revenue"] = p["revenue"].round(2)
    return p.to_dict(orient="records")


# ── 10. COHORT / WEEKDAY HEATMAP ─────────────────────────────────────────────
def weekday_revenue(df):
    order = ["Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday"]
    w = (df.groupby("weekday")
           .agg(revenue=("revenue", "sum"),
                orders=("order_id", "count"))
           .reindex(order)
           .reset_index())
    w["revenue"] = w["revenue"].round(2)
    return w.to_dict(orient="records")


# ── 11. DISCOUNT IMPACT ───────────────────────────────────────────────────────
def discount_impact(df):
    df2 = df.copy()
    df2["discount_band"] = pd.cut(
        df2["discount_pct"],
        bins=[-0.01, 0, 0.05, 0.10, 0.15, 0.25],
        labels=["No Discount", "5%", "10%", "15%", "20%+"]
    )
    d = (df2.groupby("discount_band", observed=True)
            .agg(orders=("order_id", "count"),
                 avg_revenue=("revenue", "mean"),
                 avg_review=("review_score", "mean"))
            .reset_index())
    d["avg_revenue"] = d["avg_revenue"].round(2)
    d["avg_review"]  = d["avg_review"].round(2)
    return d.to_dict(orient="records")


# ── MAIN ──────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print("🔍  Running E-Commerce Sales Analytics Pipeline...")

    output = {
        "generated_at":        datetime.now().isoformat(),
        "kpis":                compute_kpis(raw_df),
        "monthly_trend":       monthly_trend(raw_df),
        "category_performance":category_performance(raw_df),
        "state_sales":         state_sales(raw_df),
        "payment_distribution":payment_distribution(raw_df),
        "review_distribution": review_distribution(raw_df),
        "hourly_pattern":      hourly_pattern(raw_df),
        "top_products":        top_products(raw_df),
        "weekday_revenue":     weekday_revenue(raw_df),
        "discount_impact":     discount_impact(raw_df),
    }

    out_path = os.path.join(os.path.dirname(__file__), "..", "dashboard", "src", "data", "analytics.json")
    os.makedirs(os.path.dirname(out_path), exist_ok=True)

    with open(out_path, "w") as f:
        json.dump(output, f, indent=2)

    print(f"✅  analytics.json written → {os.path.abspath(out_path)}")
    print(f"\n📊  Quick Stats:")
    kpis = output["kpis"]
    print(f"    Total Revenue  : ${kpis['total_revenue']:,.2f}")
    print(f"    Total Orders   : {kpis['total_orders']:,}")
    print(f"    Avg Order Value: ${kpis['avg_order_value']:.2f}")
    print(f"    Profit Margin  : {kpis['profit_margin_pct']}%")
    print(f"    Avg Review     : {kpis['avg_review_score']} / 5")
